bigtable/client/row_range.cc

```diff
@@ -0,0 +1,446 @@
+  // Only this method can set end_key as "".
+  tmp.range_.set_end_key_open(RowRange::SAMPLE_KEY);
+  return tmp;
+}
+
+/**
+ * Infinite range has the following properties:
+ *   - 'end' value is not set.
+ *   - 'start' could be  don't cared.
+ *      The proto also has the property that, an unset start key will have
+ *      value "", inclusive.
+ *  Gives the range ["..NOT_SET).
+ *
+ *  Note: We support right-sided (or semi) inifinity, where start key can be set
+ * to any value: [x..NOT_SET); an empty end key implies semi-infinite range.
+ */
+RowRange RowRange::infinite_range() { return RowRange(); }
+
+/**
+ * Default one, right-open: [begin, end).
+ * Both start and end should be set.
+ */
+RowRange RowRange::right_open(std::string begin, std::string end) {
+  RowRange tmp;
+  tmp.range_.set_start_key_closed(std::move(begin));
+  if (!end.empty())  // (Semi-)infinite range: end = "".
+    tmp.range_.set_end_key_open(std::move(end));
+  return tmp;
+}
+
+/**
+ * open interval: (begin, end), both endpoints should be set.
+ */
+RowRange RowRange::open_interval(std::string begin, std::string end) {
+  RowRange tmp;
+  tmp.range_.set_start_key_open(std::move(begin));
+  if (!end.empty())  // (Semi-)infinite range: end = "".
+    tmp.range_.set_end_key_open(std::move(end));
+  return tmp;
+}
+
+// closed interval: [begin, end], both endpoints should be set.
+RowRange RowRange::closed_interval(std::string begin, std::string end) {
+  RowRange tmp;
+  tmp.range_.set_start_key_closed(std::move(begin));
+  if (!end.empty())  // (Semi-)infinite range: end = "".
+    tmp.range_.set_end_key_closed(std::move(end));
+  return tmp;
+}
+
+// left-open: (begin, end], both endpoints should be set.
+RowRange RowRange::left_open(std::string begin, std::string end) {
+  RowRange tmp;
+  tmp.range_.set_start_key_open(std::move(begin));
+  if (!end.empty())  // (Semi-)infinite range: end = "".
+    tmp.range_.set_end_key_closed(std::move(end));
+  return tmp;
+}
+
+/**
+ * Given any key 'prefix', the prefix range is given as the right-open interval:
+ *   ['prefix', successor('prefix')).
+ *
+ * When prefix is of the form "\xFF\xFF\xFF...\xFF", the successor is
+ * "" (empty string). In such a case, we set end key to 'prefix';
+ *
+ */
+RowRange RowRange::prefix_range(std::string prefix) {
+  RowRange tmp;
+  std::string next_prefix = Successor(prefix);
+
+  if (next_prefix.empty()) next_prefix = prefix;
+  tmp.range_.set_start_key_closed(std::move(prefix));
+  if (!next_prefix.empty())  // Infinite range for prefix="".
+    tmp.range_.set_end_key_open(std::move(next_prefix));
+  return tmp;
+}
+
+// An alias for right_open().
+RowRange RowRange::range(std::string begin, std::string end) {
+  return right_open(begin, end);
+}
+
+/*
+ * Increment the given key: gives the permutation string (in the key-space)
+ * that immediately follows the 'key'.
+ * Please note that, for a key of the form "\xFF\xFF\xFF...\xFF", the successor
+ * is "" (empty string).
```

Comment:
I do not believe this is correct.  For `"\xFF\xFF\xFF...\xFF"` the successor is `"\x00\x00\x00\...\x00\x01"`.  We are using `std::string` to represent a sequence of bytes, not a `NUL` terminated string.

Reply:
explained above. Actually, successor is "", which would be infinity. Successor  is 
 the next permutation in the key space.
For example, for "zzb", the  successor is zzc (please note the LSB, which is the last bit, not the first one); so, any key with prefix less than  zzc (say zzbbbbbbbbzzzzzz) will be less than zzc  using lexicographic comparison; the same  is true for binary  key space:  for key, 110+1 = 111, and  any string x that is lexicographically less than 111 and larger than (or equals) 110 will have prefix 110; example, 110xxxxxxxx, x is  don't care;
 For key, 111, next(111) = 000 (not 1000!); here 000 means "" (because we shrink the prefix string, discarding the trailing zeros). So, anything string (lexicographically) larger than  111  (but less than infinity) will have prefix 111.